SonarQube.Client.Tests/SonarQubeService_Lifecycle.cs

```diff
@@ -51,24 +51,47 @@ public async Task Connect_To_SonarQube_Valid_Credentials()
         }
 
         [TestMethod]
-        public void Connect_To_SonarQube_Invalid_Credentials()
+        [DataRow("http://localhost")]
+        [DataRow("https://localhost/")]
+        [DataRow("https://localhost:9000")]
+        [DataRow("http://local.sonarcloud.io/")]
+        public async Task Connect_SonarQube_IsSonarCloud_SonarQubeUrl_ReturnsFalse(string inputUrl)
         {
-            // The earliest version that supports authentication
-            SetupRequest("api/server/version", "3.3.0.0");
-            SetupRequest("api/authentication/validate", "{ \"valid\": false }");
+            var canonicalUrl = inputUrl.TrimEnd('/');
 
-            service.IsConnected.Should().BeFalse();
-            service.SonarQubeVersion.Should().BeNull();
+            // The earliest version that supports authentication
+            SetupRequest("api/server/version", "3.3.0.0", serverUrl: canonicalUrl);
+            SetupRequest("api/authentication/validate", "{ \"valid\": true }", serverUrl: canonicalUrl);
 
-            Func<Task> action = async () => await service.ConnectAsync(
-                new ConnectionInformation(new Uri("http://localhost"), "user", "pass".ToSecureString()),
+            await service.ConnectAsync(
+                new ConnectionInformation(new Uri(inputUrl), "user", "pass".ToSecureString()),
                 CancellationToken.None);
 
-            action.Should().ThrowExactly<InvalidOperationException>()
-                .And.Message.Should().Be("Invalid credentials");
+            service.IsSonarCloud.Should().BeFalse();
+        }
 
-            service.IsConnected.Should().BeFalse();
-            service.SonarQubeVersion.Should().BeNull();
+        [TestMethod]
+        [DataRow("http://sonarcloud.io")]
```

Comment:
The test cases were already tested in the lower class.
The duplication is problematic.
I think it's better to pick just the obvious SQ/SC urls here and let the underlying class be tested for all the variations.